akka-actor/src/main/scala/akka/util/Reflect.scala

```diff
@@ -81,29 +87,55 @@ private[akka] object Reflect {
       throw new IllegalArgumentException(s"$msg found on $clazz for arguments [$argClasses]")
     }
 
-    val constructor: Constructor[T] =
-      if (args.isEmpty) Try { clazz.getDeclaredConstructor() } getOrElse (null)
-      else {
-        val length = args.length
-        val candidates =
-          clazz.getDeclaredConstructors.asInstanceOf[Array[Constructor[T]]].iterator filter { c ⇒
-            val parameterTypes = c.getParameterTypes
-            parameterTypes.length == length &&
-              (parameterTypes.iterator zip args.iterator forall {
-                case (found, required) ⇒
-                  found.isInstance(required) || BoxedType(found).isInstance(required) ||
-                    (required == null && !found.isPrimitive)
-              })
+    def resolveConstructor() = {
+      val constructor: Constructor[T] =
+        if (args.isEmpty) Try { clazz.getDeclaredConstructor() } getOrElse null
+        else {
+          val length = args.length
+          val candidates =
+            clazz.getDeclaredConstructors.asInstanceOf[Array[Constructor[T]]].iterator filter { c ⇒
+              val parameterTypes = c.getParameterTypes
+              parameterTypes.length == length &&
+                (parameterTypes.iterator zip args.iterator forall {
+                  case (found, required) ⇒
+                    found.isInstance(required) || BoxedType(found).isInstance(required) ||
+                      (required == null && !found.isPrimitive)
+                })
+            }
+          if (candidates.hasNext) {
+            val cstrtr = candidates.next()
+            if (candidates.hasNext) error("multiple matching constructors")
+            else cstrtr
+          } else null
+        }
+
+      if (constructor == null) error("no matching constructor")
+      else constructor
+    }
+
+    val constructorKey = s"${clazz.getName}-${args.length}-argTypes(${
+      args.map({
+        case null    ⇒ "Any"
+        case notNull ⇒ notNull.getClass
+      }).mkString(",")
+    })"
+
+    @tailrec
+    def getConstructor(constructor: Constructor[T]): Constructor[T] = {
+      val constructorBinding = constructorBindingRef.get()
+      constructorBinding.get(constructorKey) match {
```

Comment:
Creation of that String key looks costly. It should be a case class with all parts separate.